pkg/cache/tas_flavor_snapshot.go

```diff
@@ -878,23 +990,43 @@ func belongsToRequiredDomain(leaf *leafDomain, requiredReplacementDomain utiltas
 	return strings.HasPrefix(string(utiltas.DomainID(leaf.levelValues)), string(requiredReplacementDomain))
 }
 
-func (s *TASFlavorSnapshot) fillInCountsHelper(domain *domain) int32 {
+func (s *TASFlavorSnapshot) fillInCountsHelper(domain *domain, sliceSize int32, sliceLevelIdx int, level int) (int32, int32) {
 	// logic for a leaf
 	if len(domain.children) == 0 {
-		return domain.state
+		if level == sliceLevelIdx {
+			domain.sliceState = domain.state / sliceSize
+		}
+		return domain.state, domain.sliceState
 	}
 	// logic for a parent
 	childrenCapacity := int32(0)
+	sliceCapacity := int32(0)
+
 	for _, child := range domain.children {
-		childrenCapacity += s.fillInCountsHelper(child)
+		addChildrenCapacity, addChildrenSliceCapacity := s.fillInCountsHelper(child, sliceSize, sliceLevelIdx, level+1)
+		childrenCapacity += addChildrenCapacity
+		sliceCapacity += addChildrenSliceCapacity
 	}
 	domain.state = childrenCapacity
-	return childrenCapacity
+	if level == sliceLevelIdx {
```

Comment:
```suggestion
	if level <= sliceLevelIdx {
```
To avoid filling `sliceState` with unnecessary non-zero values as described in https://github.com/kubernetes-sigs/kueue/pull/5353/files/4bfd8844c197791804491904ff032376d2f83975..904b7a02548624789bf9c8942a9f7834cbc90fef#r2142562024

wdyt?

Reply:
That's tricky part. There are two ways of calculating `sliceState`:
- Sum children's `sliceState`
- `state / sliceSize`

As an example, let's assume we have 4 topology levels: zone, block, rack, hostname. We request slices on block level. We start by assigning 0 as sliceState to all "hostname" domains (there is a special logic for leaves). Then we iterate through racks and we sum sliceStates of children - we get 0, because children have 0 everywhere, so racks also have sliceState 0. Then we iterate through blocks, but that's the slice requestes level, so we switch to 2nd way of calculating `sliceState` and we use `state / sliceSize`. Then we iterate through zones and we go back to the 1st method of summing children `sliceStates`, but children no longer have all 0 `sliceState`, so we propagate `sliceState` upwards.

If we change that "==" to "<=" we start using 2nd method of calculating `sliceState` for all domains above the slice level which will override any actual `sliceState` we can fit on the slice requested level. This will cause bugs similar to the following example:
- SliceSize = 3 pods
- Number of pods = 6 (so 2 slices)
- Slice level = host
- Rack has 3 hosts. Each host can fit 2 pods
- So rack has state = 6, but sliceState = 0, because no host can fit a single slice, so we should noFit this workload
- If we change "==" to "<=" we will calculate rack's `sliceState` = `state / size` = 2, so algorithm will say that we can fit that workload.

I'll write a test case for that as it seems it's easy to break.

Reply:
Thank you for explaining. Since this is tricky, and easy to break, as I just did without breaking any tests, then +1 for the unit test. It can be a follow up.

Reply:
Got it. I'll add that test in a follow-up then.